### PR TITLE
Implement listenable menu controller

### LINKED ISSUE #156570: [Proposal] make MenuController a ValueListenable<bool>
### Use case

Users that want to observe the `isOpen` state of a `MenuController`, for example when providing an animated icon for the child of the menu anchor.

### Proposal

The `MenuController` should be a `ValueListenable<bool>` or similar so that consumers get notified when `isOpen` changes.

Note for triage: I think I remember there being discussion about this feature in the past, but I'm not sure if someone filed a bug for it. I can't find the relevant issue at least.

### PR DESCRIPTION
**Pull request by navaronbracke**
This PR adds a new subclass of `MenuController`, that allows for listening to the menu state.
I went with a subclass to not break the existing API, as it introduces a `dispose()` method.

Fixes https://github.com/flutter/flutter/issues/156570

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by navaronbracke** [comment]
cc @nate-thegrate for additional ideas on how we could have implemented this differently.
cc @polina-c for spot checking the integration with leak tracker.

File: packages/flutter/test/material/menu_anchor_test.dart [hunk 1]
```diff
@@ -2672,6 +2672,72 @@ void main() {
...
+                    child: Text(TestMenu.mainMenu0.label),
+                  ),
+                  MenuItemButton(
+                    child: Text(TestMenu.mainMenu1.label),
+                  ),
+                  MenuItemButton(
+                    child: Text(TestMenu.mainMenu2.label),
+                  ),
+                ],
+                builder: (BuildContext context, MenuController controller, Widget? child) {
```
Comment (navaronbracke) [inline packages/flutter/test/material/menu_anchor_test.dart:2702]: I deliberately didn't change the menu anchor builder to be covariant on the menu controller.
If we ever decide that the workaround should be removed and integrated directly, then the covariant would be a point of API breakage.

File: packages/flutter/lib/src/material/menu_anchor.dart [hunk 2]
```diff
@@ -696,6 +696,73 @@ class MenuController {
   }
 }
 
+/// A [MenuController] that allows listening to its [isOpen] state.
```
Comment (navaronbracke) [inline packages/flutter/lib/src/material/menu_anchor.dart:699]: Feel free to suggest better docs.

Should we also have a `See also` that links to the `MenuController` / `MenuAnchor` ?

File: packages/flutter/lib/src/material/menu_anchor.dart [hunk 3]
```diff
@@ -696,6 +696,73 @@ class MenuController {
...
+  void dispose() {
+    if (_isDisposed) {
+      return;
+    }
+    _isDisposed = true;
+    _listeners.clear();
+
+    // TODO(polina-c): stop duplicating code across disposables
+    // https://github.com/flutter/flutter/issues/137435
+    if (kFlutterMemoryAllocationsEnabled) {
```
Comment (navaronbracke) [inline packages/flutter/lib/src/material/menu_anchor.dart:758]: I'm not sure if this should come first or last in the dispose method? (after the if-check)

**Review by nate-thegrate — COMMENTED** [review]
Unfortunately I've noticed (from my own experience along with others) that there's already some confusion about how "controller" objects work, since most—but not all—of them are listenable.

I think we'd best be able to evaluate this change once we have a long-term plan regarding how the MenuController API is going to look.

Is the plan to deprecate the non-listenable controller? Or would we want to have `MenuController` and `ListenableMenuController` coexisting from here on out?

**Comment by navaronbracke** [comment]
> Unfortunately I've noticed (from my own experience along with others) that there's already some confusion about how "controller" objects work, since most—but not all—of them are listenable.
> 
> I think we'd best be able to evaluate this change once we have a long-term plan regarding how the MenuController API is going to look.
> 
> Is the plan to deprecate the non-listenable controller? Or would we want to have `MenuController` and `ListenableMenuController` coexisting from here on out?

Yeah, the best case would be to have _only_ the listenable controller, so it "just works" with ValueListenableBuilder.
I went with the extra class to avoid introducing a breaking API change.

We could also decide to go ahead and do the breaking change, but with a migration guide? I'm fine with either.

**Comment by nate-thegrate** [comment]
Yeah this is a tricky scenario: 2 controller APIs side-by-side would probably be confusing, but I'm pretty sure making the existing controller a Listenable would be a breaking change no matter how we do it.

<br>

I took the code from the test you added and made a version where no MenuController is explicitly provided: https://dartpad.dev/?id=4be280703a9811da1c71a3f3d397a347

Unfortunately, this didn't make me feel any less conflicted.

- Providing `onOpen` / `onClose` callbacks as a workaround is a bit uglier than just having a listenable controller, but it's honestly not terrible.
- On the other hand, I know that `TextField` lets you choose whether to do state management via the listenable TextEditingController or by providing an `onChanged` callback, so perhaps the same should apply here.

**Comment by nate-thegrate** [comment]
While messing around with the MenuAnchor, I discovered something quite strange.

The [child](https://main-api.flutter.dev/flutter/material/MenuAnchor/child.html) property is documented as follows:

> Supply this child if there is a portion of the widget tree built in [builder](https://main-api.flutter.dev/flutter/material/MenuAnchor/builder.html) that doesn't depend on the controller or context supplied to the [builder](https://main-api.flutter.dev/flutter/material/MenuAnchor/builder.html). It will be more efficient, since Flutter doesn't then need to rebuild this child when those change.

After some experimenting, I discovered that the builder doesn't rebuild when the menu is opened, but does rebuild when the menu is closed. This behavior isn't documented anywhere, and I'm having trouble imagining why someone would want the builder to act this way.

<br>

I'm wondering: what if instead, we fix MenuAnchor so that the `builder` responds to both opening and closing the menu?

Then from there, we'd be able to re-evaluate the pros & cons of making the controller a listenable.

**Comment by navaronbracke** [comment]
Hmm, strange. But if we fix the builder, then we won't need this change? Then people can just check the `isOpen` flag directly, providing they use a menu controller. And since the builder would rebuild in both cases, the isOpen flag would be accurate there.

**Comment by nate-thegrate** [comment]
> But if we fix the builder, then we won't need this change? Then people can just check the `isOpen` flag directly, providing they use a menu controller.

Yes, and they wouldn't even need to explicitly create a menu controller, since it's supplied as a `builder` parameter.

If we were implementing `MenuAnchor` from scratch, I'd probably be inclined to explore a few other possibilities, but with the goal of avoiding breaking changes in mind this is probably the best way to go.

**Comment by navaronbracke** [comment]
Alright, then I'm going to close this PR. We can come up with a fix for the builder instead.

**Pull request closed without merging**